File: server/controllers/buses.js
import { MongoClient } from 'mongodb';

const CONNECTION_URL = "mongodb+srv://sudeepkarthigeyan20:sudeep2004@cluster0.6ahbt7s.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0";
const DATABASE_NAME = "Bus_details";
const COLLECTION_NAME = "coimbatore";

const client = new MongoClient(CONNECTION_URL);

export const getBuses = async (req, res) => {
    try {
        await client.connect();

        const database = client.db(DATABASE_NAME);
        const collection = database.collection(COLLECTION_NAME);

        // Fetch all buses from the database
        const allBuses = await collection.find().toArray();

        // Send the retrieved data as JSON response
        res.status(200).json(allBuses);
    } catch (error) {
        res.status(400).json({ message: error.message }); // Handle errors
    } finally {
        await client.close();
    }
};

export const locations = async (source, destination) => {
    try {
        await client.connect();

        const database = client.db(DATABASE_NAME);
        const collection = database.collection(COLLECTION_NAME);

        // Filter buses from the database based on source and destination
        const matchingBuses = await collection.find({ source: source, destination: destination }).toArray();

        return matchingBuses;
    } catch (error) {
        // Handle errors by logging or re-throwing the error
        console.error('Error fetching bus locations:', error);
        throw new Error('Failed to fetch bus locations'); // Optionally, re-throw the error
    } finally {
        await client.close();
    }
};

File: server/models/buses.py
from flask import Flask, request, jsonify
from pymongo import MongoClient

app = Flask(__name__)

# MongoDB connection details
CONNECTION_URL = "mongodb+srv://sudeepkarthigeyan20:sudeep2004@cluster0.6ahbt7s.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0"
DATABASE_NAME = "Bus_details"
COLLECTION_NAME = "coimbatore"

# Connect to MongoDB
client = MongoClient(CONNECTION_URL)
db = client[DATABASE_NAME]
collection = db[COLLECTION_NAME]

# Sample Bus Schema Representation
class Bus:
    def __init__(self, name, type, source, destination, routes):
        self.name = name
        self.type = type
        self.source = source
        self.destination = destination
        self.routes = routes

    def to_dict(self):
        return {
            "name": self.name,
            "type": self.type,
            "source": self.source,
            "destination": self.destination,
            "routes": self.routes
        }

# Endpoint to find buses by source and destination
@app.route('/buses', methods=['GET'])
def get_buses():
    try:
        source = request.args.get('source')
        destination = request.args.get('destination')

        if not source or not destination:
            return jsonify({"error": "Source and destination are required"}), 400

        # Query MongoDB for buses matching source and destination
        buses = list(collection.find({"source": source, "destination": destination}))

        # Convert ObjectId to string
        for bus in buses:
            bus['_id'] = str(bus['_id'])

        return jsonify(buses), 200

    except Exception as e:
        return jsonify({"error": str(e)}), 500

if __name__ == '__main__':
    app.run(debug=True)
